src/java.net.http/share/classes/java/net/http/HttpResponse.java

```diff
@@ -849,6 +897,67 @@ public void applyPushPromise(
             Function<HttpResponse.BodyHandler<T>,CompletableFuture<HttpResponse<T>>> acceptor
         );
 
+        /**
+         * Notification of the first occurrence of an HTTP/3 incoming push promise.
+         *
+         * Subsequent promises of the same resource (with the same PushId) are notified
+         * using {@link #notifyAdditionalPromise(HttpRequest, PushId)
+         * notifyAdditionalPromise(HttpRequest, PushId)}.
+         *
+         * <p> This method is invoked once for each push promise received, up
+         * to the point where the response body of the initiating client-sent
+         * request has been fully received.
+         *
+         * <p> A push promise is accepted by invoking the given {@code acceptor}
+         * function. The {@code acceptor} function must be passed a non-null
+         * {@code BodyHandler}, that is to be used to handle the promise's
+         * response body. The acceptor function will return a {@code
+         * CompletableFuture} that completes with the promise's response.
+         *
+         * <p> If the {@code acceptor} function is not successfully invoked,
+         * then the push promise is rejected. The {@code acceptor} function will
+         * throw an {@code IllegalStateException} if invoked more than once.
+         *
+         * @implSpec the default implementation invokes
+         * {@link #applyPushPromise(HttpRequest, HttpRequest, Function)}. This allows
+         * {@code PushPromiseHandlers} from previous releases to handle HTTP/3 push
+         * promise in a reasonable way.
+         *
+         * @param initiatingRequest the client request that resulted in the promise
+         * @param pushPromiseRequest the promised HttpRequest from the server
+         * @param pushid the PushId which can be linked to subsequent notifications
+         * @param acceptor the acceptor function that must be successfully
+         *                 invoked to accept the push promise
+         *
+         * @since TBD
+         */
+        public default void applyPushPromise(
+                HttpRequest initiatingRequest,
+                HttpRequest pushPromiseRequest,
+                PushId pushid,
```

Comment:
(Warning: I know little about HTTP, my assumptions and thus whole argument may be invalid below!)

This feels like users need to use push ids as units of synchronization on this handler, such as tracking this with a ConcurrentHashMap.

Is it possible for users to instead provide a `Function<PushId, PushPromiseHandler>` in `HttpClient::sendAsync`? I think this might make it easier for users as they no longer need to think about concurrency for different ids.

Reply:
PushId is an HTTP/3 concept. It would be strange to have a generic method (sendAsync) that you could call with any requests, but with a parameter that could only be used for HTTP/3. We don't have pushIds for HTTP/2. I don't think we would want to take this route.

Reply:
Do users need to track the push ids in a thread-safe data structure like ConcurrentHashMap in a PushPromiseHandler?

Reply:
Yes, they would need to provide a thread safe PushPromisHandler, but not in a way that would be significantly different than how they would need to do it for HTTP/2 server pushes, if they used the same handler with different concurrent requests.